### ISSUE TITLE
variable $index not found

### ISSUE THREAD

**Issue by alexsad** [davidjamesstone/superviews.js#22]
After update to version 2.0.0, the variable $index of each has replaced by variable $item ????

And now how can i do this? 
```
<div each="todo in todos"  onclick="{data.setIndex($item || $index)}">

</div>
```

I have found this manner 
```
onclick="{()=>data.setIndex($item)}"

```

but when this compile to js the result is wrong if the browser don't suport arrow functions

**Comment by davidjamesstone** [comment]
Hi there.

Yes version 1 used to wrap the `on` handlers in a function (it also called e.preventDefault for you too). I dropped this for a number of different reasons, not least because it created a new function each patch.

The reason `$value`, `$item` and `$target` were chosen was because is they made the most sense given you could be iterating over an Array, Object or Map. $index doesn't really work for Object and Map.


The function wrapper is no longer there so you have to either do it yourself inline:

```
<div each="todo in todos"  onclick="{ function (e) { data.setIndex($item || $index)} }">

</div>
```

or call a function defined on your model:

```
<div each="todo in todos"  onclick="{model.onClick}">

</div>
```

```
var model = {
  onClick: function () {...}
}

patch(el, view, model)
```


FWIW, while I _prefer_ this new way because it makes fewer assumptions, only this weekend I was thinking I may have stripped it back too much. I may reconsider the approach again (ideas are welcome!).

Do this help you for the time being?

**Comment by alexsad** [comment]
nice to me. I think if the user wants the same effect of older way it would be disponibilized as a alias with this:
```  <span click.trigger="{model.setIndexOrAnyThing($item)}"></span>  ```

**Comment by alexsad** [comment]
In a future moment could exist the "click.delagate" to pass the event to the next parent "dom" element to better memory use.

**Comment by davidjamesstone** [comment]
Hi Alex,

So I've been thinking about this over the last few days. I'm still not settled on anything, one idea I've had is:

```
<input on="{ change: fn, keyup: fn1}" onlick="{fn3}">
```

The `onclick` handler would be a straightforward DOM event handler where the `this` context is the element.

A new special `on` attribute would take an object map of handlers and would wrap the functions, like it did in superviews@1.0.0, additionally calling e.preventDefault().

Delegated events would be nice too, although would probably require a third party lib e.g. [ftdomdelegate](https://github.com/ftlabs/ftdomdelegate)

**Comment by alexsad** [comment]
if this will be translate to 
```
<input onchange="function(e){e.preventDefault();fn($item) }">
```
Good to me,
but how we could do :  

```
<input on="{ change: (fn($item);data.index=$item)  }" >
```
?

becouse in the older version we do this : 
``` <form onsubmit="{data.sendItem();data.text=''}"> ```

**Comment by davidjamesstone** [comment]
You're right. The 'on' way I described would not give us what we used to have.

I'm tempted to just go back to how it was (but without the e.preventDefault() call):

```js
<input onchange="{data.value = this.value}">
```

Compiles to

```js
<input onchange="function ($event) { data.value = this.value }">
```

Would you be happy with this?

**Comment by alexsad** [comment]
for me is perfect, because I do onclick = "$event.preventDefault (); data.value=this.value;fnsetindex($item)" when you need to make the preventDefault. But if that is harming you or please implementation we will be discussing but.

**Comment by davidjamesstone** [comment]
OK - I've made the changes and released as 3.0.0 on npm.

We can continue this discussion but for now I think reverting to the old way (without e.preventDefault()) is the best thing.

**Comment by alexsad** [comment]
great thanks.
